### ISSUE TITLE
Support L7 Host Policies

### ISSUE THREAD

**Issue by pchaigno** [cilium/cilium#13599]
The host firewall currently doesn't support L7 rules in host policies. Such rules are [rejected during the validation of policies](https://github.com/cilium/cilium/blob/9d3df3c47d39a123eadd99332c0ebcdbda1732db/pkg/policy/api/rule_validation.go#L76). The main reason is the lack of support in the datapath implementation of the host firewall.

I haven't looked at what is needed exactly, but here's how I would proceed:
- [ ] Remove the validation check mentioned above. 
- [ ] Write a basic L7 host policy check and verify that the BPF maps have the proper content (i.e., with redirect port). That part should already work.
- [ ] Extend the implementation in `host_firewall.h` to support redirects to the L7 proxy in the same way as we already do for pod policies (`bpf_lxc.c`).

**Comment by joestringer** [comment]
Note, this could get a bit complicated because the original outbound connection from the original application and the connection from the proxy both originate from the same network namespace, which is different compared to the standard pod -> proxy -> out path.

I would highly suggest extending the `examples/kubernetes/connectivity-check/` definitions to add a new set of pods that perform requests from/to a hostnetwork pod, with an L7 policy as well. You could test the various paths - host traffic subject to L7 egress policy, likewise L7 ingress policy, also both ingress+egress L7 policy for a path (eg host-to-host). The latter should furthermore be tested both within the node and across nodes. This could act as a test harness to investigate each of the possible paths and work on them.

**Comment by pchaigno** [comment]
> because the original outbound connection from the original application and the connection from the proxy both originate from the same network namespace, which is different compared to the standard pod -> proxy -> out path.

@joestringer Don't we have packet marks `MARK_MAGIC_PROXY_INGRESS` and `MARK_MAGIC_PROXY_EGRESS` to differentiate packets egressing the proxy from other packets egressing the host?

**Comment by joestringer** [comment]
True, yes. As long as we take these paths into account with whatever we write, we should be good.

**Comment by viveksahu26** [comment]
Hey, @pchaigno, I know the basics of Ingress(Nginx), how it works, its rule-book. I also have a knowledge of CNI and how it adds firewall using Ingress to make Network Policy.  I have golang basic to intermediate level. Now, coming to the issue, how to "Remove the validation check".  Please guide me.

**Comment by pchaigno** [comment]
Hi @viveksahu26 :wave: 
Please use the #gsoc channel on [our Slack](https://cilium.herokuapp.com/) to introduce yourself and ask questions. There, we'll be happy to answer any questions you have :smiley_cat:

**Comment by mllab-nl** [comment]
@pchaigno, @julianwiedmann is there a chance this issue will get more attention in the foreseeable future?

**Comment by pchaigno** [comment]
> is there a chance this issue will get more attention in the foreseeable future?

That's hard to predict. Someone from the community would have to have strong enough interest that they are ready to contribute this work (or to outsource it).

>  Is enabling of FQDN rules in Host policies tracked by this issue as well ?

Yes. L7 includes DNS as both types of policies require some sort of proxy redirection.

**Comment by qmonnet** [comment]
- https://github.com/cilium/cilium/pull/34024
